compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -550,30 +553,42 @@ fn collate_raw_dylibs(used_libraries: &[NativeLib]) -> Vec<(String, Vec<DllImpor
             } else {
                 format!("{}.dll", name)
             };
-            dylib_table
-                .entry(name)
-                .or_default()
-                .extend(lib.dll_imports.iter().map(|import| import.name));
+            dylib_table.entry(name).or_default().extend(lib.dll_imports.iter().cloned());
         }
     }
 
-    // FIXME: when we add support for ordinals, fix this to propagate ordinals.  Also figure out
-    // what we should do if we have two DllImport values with the same name but different
-    // ordinals.
+    // Check for multiple imports with the same name but different calling conventions or
+    // (when relevant) argument list sizes.  Rustc only signals an error for this if the
+    // declarations are at the same scope level; if one shadows the other, we only get a lint
+    // warning.
+    for (library, imports) in &dylib_table {
+        let mut import_table: FxHashMap<Symbol, &DllImport> = FxHashMap::default();
+        for import in imports {
+            if let Some(old_import) = import_table.insert(import.name, &import) {
+                if import != old_import {
```

Comment:
`DllImport` contains a `Span` now, so you cannot use derived `PartialEq`/`Hash` anymore, spans are almost always going to be different.
`import_table` should probably keep only `DllCallingConvention`s instead of full `DllImport`s.

Reply:
Ok, it looks like I'm sort of stuck here.  As far as I can determine, something in the query system requires `NativeLib` to implement `HashStable`.  (I'm sort of guessing here because the error happens somewhere in a big stack of macro invocations, and I'm having trouble figuring out where in this stack this requirement is imposed.)  Therefore, adding `Span` information to `NativeLib` (even indirectly, as I have) doesn't seem to be an option.

The `Span` is only necessary so I could give some sort of source location information in the error message that we signal if we encounter multiple declarations of the same external function with inconsistent calling conventions.  Do we want to settle for issuing a diagnostic without location info?  (I'm not wild about this choice.)  Alternatively, is there a better place in the compiler to detect this case and issue the error?  I would expect that a big difficulty with doing this before we've created the `NativeLib` structs is that we don't have easy access to information about all of the external declarations in one place.

> `import_table` should probably keep only `DllCallingConvention`s instead of full `DllImport`s.

I think this gets into a question about how much future proofing we should be doing here.  The RFC does call for eventual support of linking by ordinal instead of by name, and `DllImport` contains the `ordinal` field to support this in the future.  At the moment, nobody ever reads that field.  Going forward, I have no idea if/when I'm going to get around to implementing that specific bit of functionality; our primary concern is to make interoperability with the Windows API easier, and support for ordinals isn't very important for that.  I'm starting to wonder if I should just drop the field (but otherwise leave `DllImport` intact) rather than including only half-hearted support for ordinals.

Reply:
>something in the query system requires `NativeLib` to implement `HashStable`

Derived `Hash` implementation (including `Span`) works ok for the purposes of query/incremental systems, it'll just add some dependency on spans, so it can be kept.
The important thing is to not use for logic like duplicate detection here.

>I think this gets into a question about how much future proofing we should be doing here. The RFC does call for eventual support of linking by ordinal instead of by name

Then `FxHashMap<Symbol, DllCallingConvention>` will turn into `FxHashMap<Symbol, (DllCallingConvention, Option<u16>)>`.